ir/src/boundary_stmts.rs

```diff
@@ -0,0 +1,201 @@
+use crate::{BoundaryConstraintsMap, TraceSegment};
+
+use super::{BTreeMap, BoundaryExpr, IdentifierType, SemanticError, SymbolTable};
+use parser::ast::{self, BoundaryStmt};
+
+// BOUNDARY CONSTRAINTS
+// ================================================================================================
+
+/// A struct containing all of the boundary constraints to be applied at each of the 2 allowed
+/// boundaries (first row and last row). The constraints are stored in a vector of
+/// [SegmentBoundaryConstraints] where each index of the vector corresponds to the trace segment
+/// of the column the boundary constraint applies to. For ease of code generation and evaluation,
+/// constraints are sorted into maps by the boundary. This also simplifies ensuring that there are
+/// no conflicting constraints sharing a boundary and column index.
+#[derive(Default, Debug)]
+pub(crate) struct BoundaryStmts {
+    boundary_constraints: Vec<SegmentBoundaryConstraints>,
+}
+
+/// A struct containing boundary constraints applied to columns in a segment at the first and last
+/// rows.
+#[derive(Default, Debug, Clone)]
+struct SegmentBoundaryConstraints {
+    first: BoundaryConstraintsMap,
+    last: BoundaryConstraintsMap,
+}
+
+impl SegmentBoundaryConstraints {
+    fn new() -> Self {
+        Self {
+            first: BTreeMap::new(),
+            last: BTreeMap::new(),
+        }
+    }
+
+    fn first(&self) -> &BoundaryConstraintsMap {
+        &self.first
+    }
+
+    fn last(&self) -> &BoundaryConstraintsMap {
+        &self.last
+    }
+
+    fn first_mut(&mut self) -> &mut BoundaryConstraintsMap {
+        &mut self.first
+    }
+
+    fn last_mut(&mut self) -> &mut BoundaryConstraintsMap {
+        &mut self.last
+    }
+}
+
+impl BoundaryStmts {
+    /// Creates a new [BoundaryStmts] instance with the specified number of trace segments.
+    pub fn new(num_trace_segments: usize) -> Self {
+        Self {
+            boundary_constraints: vec![SegmentBoundaryConstraints::new(); num_trace_segments],
+        }
+    }
+
+    // --- ACCESSORS ------------------------------------------------------------------------------
+
+    /// Returns the number of boundary constraints for the specified trace segment.
+    pub fn num_boundary_constraints(&self, trace_segment: TraceSegment) -> usize {
+        if self.boundary_constraints.len() <= trace_segment.into() {
+            0
+        } else {
+            self.boundary_constraints[trace_segment as usize]
+                .first
+                .len()
+                + self.boundary_constraints[trace_segment as usize].last.len()
+        }
```

Comment:
nit: Since it's used 3 times, I would probably convert `trace_segment` to `usize` just once at the top, e.g. `let trace_segment = trace_segment as usize`